File: filters/src/slipbox.ts
import { strict as assert } from 'assert'
import * as process from 'process'

import Database from 'better-sqlite3'

import { warning } from './log.js'
import { parentAlias } from './utils.js'

export class Slipbox {
  db: Database.Database
  insert: { [key: string]: Database.Statement }
  update: { [key: string]: Database.Statement }
  select: { [key: string]: Database.Statement }
  notes: { [id: string]: { title: string, filename: string } } = {}
  citations: { [id: string]: Set<string> } = {}
  aliases: { [alias: string]: { id: string, owner: string } } = {}
  sequences: Array<[string, string]> = []
  links: Array<{ tag: string, src: string, dest: string, description: string }> = []
  tags: Array<[string, string]> = []
  references: Array<[string, string]> = []
  errors: {
    hasEmptyLink: Set<string>
  } = { hasEmptyLink: new Set() }

  constructor () {
    const path = process.env.SLIPBOX_DB || 'slipbox.db'
    this.db = new Database(path, { fileMustExist: true })

    const db = this.db
    this.insert = {
      alias: db.prepare('INSERT OR IGNORE INTO Aliases (id, owner, alias) VALUES (?, ?, ?)'),
      bib: db.prepare('INSERT OR IGNORE INTO Bibliography (key, text) VALUES (?, ?)'),
      cite: db.prepare('INSERT INTO Citations (note, reference) VALUES (?, ?)'),
      file: db.prepare('INSERT OR IGNORE INTO Files (filename) VALUES (?)'),
      link: db.prepare('INSERT INTO Links (src, dest, annotation) VALUES (?, ?, ?)'),
      note: db.prepare('INSERT INTO Notes (id, title, filename) VALUES (?, ?, ?)'),
      sequence: db.prepare('INSERT OR IGNORE INTO Sequences (prev, next) VALUES (?, ?)'),
      tag: db.prepare('INSERT OR IGNORE INTO Tags (id, tag) VALUES (?, ?)')
    }
    this.update = {
      bib: db.prepare('UPDATE Bibliography SET text = ? WHERE key = ?')
    }
    this.select = {
      note: db.prepare('SELECT id, title, filename FROM Notes WHERE id = ?'),
      noteFromAlias: db.prepare(
        'SELECT id, title, filename FROM Notes JOIN Aliases USING (id) WHERE alias = ?'
      )
    }
  }

  saveNote (id: string, title: string, filename: string) {
    const existing = this.notes[id]
    if (existing == null) {
      this.notes[id] = { title, filename }
    } else {
      warning([
        `Duplicate ID: ${id}.`,
        `Could not insert note '${title}'.`,
        `Note '${existing.title}' already uses the ID.`
      ])
    }
  }

  saveCite (id: string, ref: string) {
    const rec = new Set(this.citations[id] || [])
    rec.add('ref-' + ref)
    this.citations[id] = rec
  }

  saveAlias (alias: string, id: string, owner: string) {
    const existing = this.aliases[alias]
    if (existing && existing.id !== id) {
      warning([
        `Duplicate alias definition for '${alias}' used by note ${existing.id}.`,
        `It will not be used as an alias for note ${id}.`
      ])
    } else {
      this.aliases[alias] = { id, owner }
      const prev = parentAlias(alias)
      if (prev != null) {
        this.sequences.push([prev, alias])
      }
    }
  }

  saveLink (src: string, dest: string, description: string) {
    this.links.push({ tag: 'direct', src, dest, description })
  }

  saveTag (id: string, tag: string) {
    this.tags.push([id, tag])
  }

  saveNotes () {
    const insertMany = this.db.transaction(() => {
      for (const note of Object.entries(this.notes)) {
        const [id, { title, filename }] = note
        this.insert.file.run(filename)
        try {
          this.insert.note.run([id, title, filename])
        } catch (error) {
          const existing = this.select.note.get(id)
          assert(existing)
          const messages = [
            `Duplicate ID: ${id}.`,
            `Could not insert note '${title}'.`,
            `Note '${existing.title}' already uses the ID.`,
            `See '${filename}' or '${existing.filename}'.`
          ]
          warning(messages)
        }
      }
    })
    insertMany()
    this.notes = {} // ???
  }

  saveCitations () {
    const insertMany = this.db.transaction(() => {
      for (const [id, _cites] of Object.entries(this.citations)) {
        _cites.forEach(cite => {
          this.insert.bib.run([cite, ''])
          this.insert.cite.run([id, cite])
        })
      }
    })
    insertMany()
    this.citations = {} // ???
  }

  saveAliases () {
    const insertMany = this.db.transaction(() => {
      for (const [alias, { id, owner }] of Object.entries(this.aliases)) {
        this.insert.alias.run([owner, owner, String(owner)])
        this.insert.alias.run([id, owner, alias])
      }
    })
    insertMany()
    this.aliases = {} // ???
  }

  saveSequences () {
    const insertMany = this.db.transaction(() => {
      for (const [prev, next] of this.sequences) {
        assert(prev && next)
        try {
          this.insert.sequence.run([prev, next])
        } catch (error) {
          // NOTE assume prev is the missing alias
          const existing = this.select.noteFromAlias.get(next)
          assert(existing)
          warning([
            `Missing note alias: '${prev}'.`,
            `Note ${existing.id} with alias '${next}' will be unreachable.`
          ])
          // TODO What if note with parent alias with be added in the
          // next batch (i.e. has different file extension)?
        }
      }
    })
    insertMany()
    this.sequences = [] // ???
  }

  saveLinks () {
    // TODO handle error if note has duplicate links to a note
    const insertMany = this.db.transaction(() => {
      for (const { tag, src, dest, description } of this.links) {
        assert(tag === 'direct')
        this.insert.link.run([src, dest, description])
      }
    })
    insertMany()
    this.links = [] // ???
  }

  saveTags () {
    const insertMany = this.db.transaction(() => {
      for (const [id, tag] of this.tags) {
        this.insert.tag.run([id, tag])
      }
    })
    insertMany()
    this.tags = [] // ???
  }

  saveReferences (references: Array<[string, string]>) {
    const insertMany = this.db.transaction(() => {
      for (const [key, text] of references) {
        this.update.bib.run([text, key])
      }
    })
    insertMany()
  }

  hasEmptyLink (id: string) {
    this.errors.hasEmptyLink.add(id)
  }

  checkEmptyLinks () {
    const ids = Array.from(this.errors.hasEmptyLink).map(x => `'${x}'`).join(', ')
    const result = this.db.prepare(`
      SELECT id, title, filename
      FROM Notes
      WHERE id IN (${ids})
      ORDER BY id
    `).all()
    const messages = ['The notes below contain links with an empty target.']
    for (const { id, title, filename } of result) {
      messages.push(`${id}. ${title} in '${filename}'.`)
    }
    if (messages.length > 1) {
      warning(messages)
    }
  }
}

File: filters/src/slipbox.lua
local csv = require "src.csv"
local errors = require "src.errors"
local utils = require "src.utils"

local SlipBox = {}
function SlipBox:new()
  self.__index = self
  return setmetatable({
    files = {},
    notes = {},
    tags = {},
    links = {},
    citations = {},
    images = {},
    bibliography = {},
    invalid = {
      has_empty_link_target = {},
    },
  }, self)
end

function SlipBox:save_file(filename, hash)
  assert(type(filename) == "string")
  assert(type(hash) == "string")
  assert(self.files[filename] == nil or self.files[filename] == hash)
  self.files[filename] = hash
end

function SlipBox:save_citation(id, citation)
  -- Save citation from note id (number).
  assert(type(id) == "number")
  local citations = self.citations[id] or {}
  citations[citation] = true
  self.citations[id] = citations
end

function SlipBox:save_image(id, filename)
  assert(type(id) == "number")
  assert(type(filename) == "string")

  local image = self.images[filename] or {
    id = id,
    filename = filename,
    notes = {},
  }
  self.images[filename] = image
  image.notes[id] = true
end

function SlipBox:save_reference(id, html)
  -- Save reference into slipbox.
  assert(type(id) == "string")
  assert(type(html) == "string")
  assert(id ~= "")
  assert(html ~= "")
  self.bibliography[id] = html
end

function SlipBox:save_note(id, title, filename)
  -- Save note into slipbox.
  -- Returns false if a note with the same ID already exists, true otherwise.
  -- Logs errors if any.
  assert(type(id) == "number")
  assert(type(title) == "string")
  assert(type(filename) == "string")
  assert(title ~= "")
  assert(filename ~= "")

  local note = self.notes[id]
  if note and (note.title ~= title or note.filename ~= filename) then
    errors.duplicate_note_id(id, {note, {title = title, filename = filename}})
    return false
  end
  self.notes[id] = {title = title, filename = filename}
  return true
end

function SlipBox:save_tag(id, tag)
  assert(type(id) == "number")
  assert(type(tag) == "string")
  assert(tag ~= "")

  local tags = self.tags[id] or {}
  tags[tag] = true
  self.tags[id] = tags
end

function SlipBox:save_link(link)
  if link and link.src then
    local links = self.links[link.src] or {}
    table.insert(links, link)
    self.links[link.src] = links
  end
end

local function notes_to_csv(notes)
  -- Generate CSV data from slipbox notes.
  local w = csv.Writer:new{"id", "title", "filename"}
  for id, note in pairs(notes) do
    if note.filename then
      w:write{id, note.title, note.filename}
    end
  end
  return w.data
end

local function tags_to_csv(all_tags)
  -- Create CSV data from tags in slipbox.
  local w = csv.Writer:new{"tag", "id"}
  for id, tags in pairs(all_tags) do
    for tag, _ in pairs(tags) do
      w:write{tag, id}
    end
  end
  return w.data
end

local function links_to_csv(links)
  -- Create CSV data from direct links in slipbox.
  local w = csv.Writer:new{"src", "dest", "direction"}
  for src, dests in pairs(links) do
    for _, dest in ipairs(dests) do
      w:write{src, dest.dest, dest.direction}
    end
  end
  return w.data
end

local function bibliography_to_csv(refs)
  local w = csv.Writer:new{"key", "html"}
  for ref, html in pairs(refs) do
    w:write{ref, html}
  end
  return w.data
end

local function files_to_csv(files)
  local w = csv.Writer:new{"filename", "hash"}
  for filename, hash in pairs(files) do
    w:write{filename, hash}
  end
  return w.data
end

local function citations_to_csv(citations)
  local w = csv.Writer:new{"note", "reference"}
  for id, cites in pairs(citations) do
    for cite in pairs(cites) do
      w:write{id, cite}
    end
  end
  return w.data
end

local function images_to_csv(images)
  local w = csv.Writer:new{"filename"}
  for filename in pairs(images) do
    w:write{filename}
  end
  return w.data
end

local function image_links_to_csv(images)
  local w = csv.Writer:new{"note", "image"}
  for filename, image in pairs(images) do
    for note in pairs(image.notes) do
      w:write{note, filename}
    end
  end
  return w.data
end

function SlipBox:write_data()
  -- Create CSV data to files.
  local write = utils.write_text
  write("files.csv", files_to_csv(self.files))
  write("notes.csv", notes_to_csv(self.notes))
  write("tags.csv", tags_to_csv(self.tags))
  write("links.csv", links_to_csv(self.links))
  write("images.csv", images_to_csv(self.images))
  write("image_links.csv", image_links_to_csv(self.images))
  write("bibliography.csv", bibliography_to_csv(self.bibliography))
  write("citations.csv", citations_to_csv(self.citations))
end

return {
  SlipBox = SlipBox,
}
